include/avif/internal.h

```diff
@@ -283,10 +291,22 @@ typedef avifBool (*avifCodecGetNextImageFunc)(struct avifCodec * codec,
 // encoded and EncodeFinish is called, the number of samples emitted must match the number of submitted frames.
 // avifCodecEncodeImageFunc may return AVIF_RESULT_UNKNOWN_ERROR to automatically emit the appropriate
 // AVIF_RESULT_ENCODE_COLOR_FAILED or AVIF_RESULT_ENCODE_ALPHA_FAILED depending on the alpha argument.
+// avifCodecEncodeImageFunc should use tileRowsLog2 and tileColsLog2 instead of
+// encoder->tileRowsLog2, encoder->tileColsLog2, and encoder->autoTiling. The caller of
+// avifCodecEncodeImageFunc is responsible for automatic tiling if encoder->autoTiling is set to
+// AVIF_TRUE. The actual tiling values are passed to avifCodecEncodeImageFunc as parameters.
+//
+// Note: The caller of avifCodecEncodeImageFunc always passes encoder->data->tileRowsLog2 and
+// encoder->data->tileColsLog2 as the tileRowsLog2 and tileColsLog2 arguments. Because
+// encoder->data is of a struct type defined in src/write.c, avifCodecEncodeImageFunc cannot
+// dereference encoder->data and has to receive encoder->data->tileRowsLog2 and
+// encoder->data->tileColsLog2 via function parameters.
 typedef avifResult (*avifCodecEncodeImageFunc)(struct avifCodec * codec,
                                                avifEncoder * encoder,
                                                const avifImage * image,
                                                avifBool alpha,
+                                               int tileRowsLog2,
```

Comment:
I wonder if it's better to pass these values using `avifCodec` instead of as separated parameters? There's already `codec.allLayers` for decoder side whose value depends on `decoder.allowProgressive` field.

Reply:
Yuan: Thank you for the suggestion. After finding that src/codec_*.c cannot dereference `encoder->data`, I looked into this option and several other options. They all have their pros and cons. My preference was to define a new struct to collect the existing `encoderChanges` parameter and the new `tileRowsLog2` and `tileColsLog2` parameters, but in the end I decided to imitate `encoderChanges` and just add `tileRowsLog2` and `tileColsLog2` as parameters.

I can also share my thought process behind this pull request. There were three design decisions.

1. Whether to handle automatic tiling in src/write.c or in the individual src/codec_*.c files.
2. Where to store the actual tiling values (we need to store them so we can detect changes)
3. If src/write.c handles automatic tiling, how to pass the actual tiling values to src/codec_*.c files.

I decided to handle automatic tiling in src/write.c to avoid duplicating code. But not much code would be duplicated if we were to handle automatic tiling in src/codec_*.c.

It's tempting to store the actual tiling values in `encoder->tileRowsLog2` and `encoder->tileColsLog2` directly. It would allow libavif clients to know the actual tiling values used by the encoder. But I think it'd be strange to let `encoder` become an in/out parameter. So I didn't do that.

Reply:
Wan-Teh: Thanks for your detail explanation. I also prefer using a new struct to collect them, especially if we are going to add more parameters like these. I think `avifCodec` may be used for this, but it's not a perfect fit either.

Reply:
Yes. Some avifCodecs such as rav1e and SVT-AV1 are encoders only, so it isn't ideal to put decoder-specific fields such as `operatingPoint` and `allLayers` in the `avifCodec` struct.